### PR TITLE
lsp: make tagstack smarter motion-wise

### PR DESCRIPTION
**Pull request by gbrlsnchs**
From the commit description:
```
This commit prevents two things regarding the tagstack and jumping to
locations:

  - Pushing the same item twice in a row

  - Pushing an item where the destination is the same as the source

Both prevent having to press CTRL-T additional times just to pop items
that don't make the cursor move.
```

### CODE REVIEW

File: runtime/lua/vim/lsp/util.lua [hunk 1]
```diff
@@ -423,27 +423,44 @@ function M.make_floating_popup_options(width, height, opts)
...
-
-  --- Jump to new location
-  api.nvim_set_current_buf(bufnr)
-  api.nvim_buf_set_option(0, 'buflisted', true)
+  local item = {tagname=vim.fn.expand('<cword>'), from=from}
   local range = location.range or location.targetSelectionRange
   local row = range.start.line
   local col = range.start.character
-  local line = api.nvim_buf_get_lines(0, row, row+1, true)[1]
-  col = vim.str_byteindex(line, col)
```
Comment (clason) [inline runtime/lua/vim/lsp/util.lua:446]: This removes a correction that is necessary to get correct column information when the line contains non-ASCII characters before `col` (UTF-16 encoding...)

**Comment by clason** [comment]
Maybe it would make sense to factor this out as a `add_tagstack` function instead? (seems a bit more expressive than `should_add`, and keeps `jump_to_location` cleaner)

File: runtime/lua/vim/lsp/util.lua (hunk 1 above)
Comment (gbrlsnchs) [reply runtime/lua/vim/lsp/util.lua:446]: Oh, I see. Thanks for pointing that out. I'll revert this.

**Comment by gbrlsnchs** [comment]
> Maybe it would make sense to factor this out as a `add_tagstack` function instead? (seems a bit more expressive than `should_add`, and keeps `jump_to_location` cleaner)

Makes sense to me. I'll convert this PR back to draft so I can work on this again. Thanks for the review!

**Comment by gbrlsnchs** [comment]
@clason I have made some changes to organize code better. What do you think now?

**Review by h-michael — COMMENTED** [review]
Great! Detailed commit messages and code comments are very useful.

I'm sorry because it doesn't exist originally, would you add a test for jump_to_location to "test/functional/lua/vim_spec.lua"?
If you don't know how to write well, we can support you.

File: runtime/lua/vim/lsp/util.lua [hunk 2]
```diff
@@ -423,27 +423,58 @@ function M.make_floating_popup_options(width, height, opts)
...
+  return false
+end
+
+local function add_tagstack(item)
+  -- Check whether item is not already at the top of the tagstack.
+  if not is_new_item(item) then return end
+  local winid = vim.fn.win_getid()
+  local items = {item}
+  vim.fn.settagstack(winid, {items=items}, 't')
+end
```
[the remark this answers is not in the document]
Comment (clason) [reply runtime/lua/vim/lsp/util.lua:449]: I disagree with the first part (no surprise); If the goal of nvim LSP is to have a bare-bones but hackable implementation, then functions should be as small and composable as possible -- otherwise "hacking" means "copying the whole whole file and changing a line".

I do agree about the more expressive name, though!

**Comment by clason** [comment]
> @clason I have made some changes to organize code better. What do you think now?

Looks good! I'm just wondering whether it's necessary (or useful) to separate `is_new_item` out from `add_tagstack` (my above comment notwithstanding).

File: runtime/lua/vim/lsp/util.lua [hunk 3]
```diff
@@ -423,27 +423,58 @@ function M.make_floating_popup_options(width, height, opts)
...
+  -- Check if the item at the top of the tagstack is exactly the
+  -- same as the one we want to push.
+  if top_item.tagname ~= item.tagname then return true end
+  for i, v in ipairs(top_item.from) do
+    if v ~= item.from[i] then return true end
+  end
+  return false
+end
+
+local function add_tagstack(item)
```
Comment (justinmk) [inline runtime/lua/vim/lsp/util.lua:443]: should be private :
```suggestion
local function _add_tagstack(item)
```

File: runtime/lua/vim/lsp/util.lua [hunk 4]
```diff
@@ -423,27 +423,58 @@ function M.make_floating_popup_options(width, height, opts)
   }
 end
 
+local function is_new_item(item)
```
Comment (justinmk) [inline runtime/lua/vim/lsp/util.lua:426]: private
```suggestion
local function _is_new_item(item)
```

File: runtime/lua/vim/lsp/util.lua (hunk 3 above)
Comment (gbrlsnchs) [reply runtime/lua/vim/lsp/util.lua:443]: Thanks for the tips, I'm new to Lua, so it's nice to learn from you all.

**Comment by gbrlsnchs** [comment]
> I'm sorry because it doesn't exist originally, would you add a test for jump_to_location to "test/functional/lua/vim_spec.lua"?

I checked that file and it only contains tests for `vim.*` functions. Wouldn't my tests fit better in `test/functional/plugin/lsp_spec.lua`?

> If you don't know how to write well, we can support you.

Since this is my first experience with Lua (and contributing to Neovim), I have no idea about:
- How to run Lua tests locally
- How to prepare the test (since I need to jump to a location, are editor APIs mocked somehow? how to I simulate a buffer, my position, etc)
- What should I expect as success (final cursor position? final tagstack items?)

Is there a guide for creating Lua tests? I really have no idea how tests work in this context. If there is a manual or something, it'll already help a lot. Thanks!

**Comment by h-michael** [comment]
@gbrlsnchs 

> How to run Lua tests locally

[This](https://github.com/neovim/neovim/tree/master/test) is documentation about test.

> How to prepare the test (since I need to jump to a location, are editor APIs mocked somehow? 

I think that you should not mock API for writing test.
There are many helper for test. (See test/functional/helper)

> What should I expect as success (final cursor position? final tagstack items?)

I think that it should check tagstack items.

**Comment by gbrlsnchs** [comment]
@h-michael I have just finished adding tests and code is now ready for review (I organized it as requested) . Thankfully someone wrote some base tests for "jump to location" and I simply added test cases for tagstack there using the helpers provided. 😸

**Comment by h-michael** [comment]
@gbrlsnchs Thanks! I'll review this tonight or tomorrow :)

**Comment by gbrlsnchs** [comment]
By the way, what is the reason for CI on OpenBSD to fail and not on FreeBSD though?

**Review by h-michael — APPROVED** [review]
Looks good to me. @clason How do you thik?

File: runtime/lua/vim/lsp/util.lua [hunk 5]
```diff
@@ -525,26 +525,57 @@ function M.make_floating_popup_options(width, height, opts)
...
-  vim.cmd "normal! m'"
-
-  -- Push a new item into tagstack
-  local from = {vim.fn.bufnr('%'), vim.fn.line('.'), vim.fn.col('.'), 0}
-  local items = {{tagname=vim.fn.expand('<cword>'), from=from}}
-  vim.fn.settagstack(vim.fn.win_getid(), {items=items}, 't')
 
   --- Jump to new location (adjusting for UTF-16 encoding of characters)
+  local bufnr = vim.uri_to_bufnr(uri)
   api.nvim_set_current_buf(bufnr)
```
Comment (tjdevries) [inline runtime/lua/vim/lsp/util.lua:560]: It seems weird that this jumps to the new buffer before checking where it came from.

Am I wrong there?

File: test/functional/plugin/lsp_spec.lua [hunk 6]
```diff
@@ -1378,6 +1378,33 @@ describe('LSP', function()
...
+      eq(2, stack.length)
+      eq({target_bufnr, 1, 10, 0}, stack.items[2].from)
+    end)
+
+    it('prevents pushing to tagstack when motionless', function()
+      -- Jump to the same location the cursor is parked.
+      jump(location(0, 0, 0, 0))
+      local stack = exec_lua[[return vim.fn.gettagstack()]]
+      eq(0, stack.length)
+    end)
```
Comment (tjdevries) [inline test/functional/plugin/lsp_spec.lua:1407]: Do you think you could add a test where we jump to a different buffer? I'm not sure we're doing that here. That would be good anyway, since it seems easy for us to mess up in the future.

Comment (tjdevries) [reply test/functional/plugin/lsp_spec.lua:1407]: Like jump from buffer 1, to buffer 2, to buffer 1, for example.

Or, jump from buffer 1 to buffer 2, but in the same cursor location. That might be a good test as well.

File: runtime/lua/vim/lsp/util.lua [hunk 7]
```diff
@@ -525,26 +525,57 @@ function M.make_floating_popup_options(width, height, opts)
...
+    local top_item = stack.items[stack.curidx-1]
+
+    -- Check if the item at the top of the tagstack is exactly the
+    -- same as the one we want to push.
+    if top_item.tagname ~= new_item.tagname then return true end
+    for i, v in ipairs(top_item.from) do
+      if v ~= new_item.from[i] then return true end
+    end
+    return false
+  end
```
Comment (tjdevries) [inline runtime/lua/vim/lsp/util.lua:544]: Why is this a function here?

You could either define it outside of the above function or just make this inlined, right? I don't currently see why it's written this way.

File: runtime/lua/vim/lsp/util.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (tjdevries) [reply runtime/lua/vim/lsp/util.lua]: I don't agree with this 😁 I think it's fine to write a function outside of the function, for clarity sake. This function gets redefined every time we call the function, which I don't think makes sense.

I left a comment to either just inline it, or we can move it back out. The name could be clearer, or added w/in a `do` block to hide it's scope, but I am not a big fan of this style choice. Of course, if I'm outnumbered, I will acquiesce 😆

File: runtime/lua/vim/lsp/util.lua (hunk 7 above)
Comment (tjdevries) [reply runtime/lua/vim/lsp/util.lua:544]: (I just now see the comment you left earlier @h-michael . I have left a comment there now as well).

**Comment by tjdevries** [comment]
> By the way, what is the reason for CI on OpenBSD to fail and not on FreeBSD though?

We have a test that fails there. Some other people are currently working on fixing that. Don't worry about it :)

File: runtime/lua/vim/lsp/util.lua (hunk 7 above)
Comment (gbrlsnchs) [reply runtime/lua/vim/lsp/util.lua:544]: I don't know Lua well enough to justify this, unfortunately. There is overhead by declaring it like I did, right? Or does it get inlined somehow?

File: runtime/lua/vim/lsp/util.lua (hunk 5 above)
Comment (gbrlsnchs) [reply runtime/lua/vim/lsp/util.lua:560]: UX-wise, it should act as a no-op. The tagstack check is done afterwards. I'm not sure about any possible overheads involving it though.

File: test/functional/plugin/lsp_spec.lua (hunk 6 above)
Comment (gbrlsnchs) [reply test/functional/plugin/lsp_spec.lua:1434]: Makes sense, more tests the better!

File: runtime/lua/vim/lsp/util.lua (hunk 5 above)
Comment (gbrlsnchs) [reply runtime/lua/vim/lsp/util.lua:560]: Sorry, @tjdevries, you were totally right about this. This was not working, actually, when jumping from one buffer to another. Thank you very much for hinting me about this.

I have redone everything, including tests. They are now documented. :cat:

File: test/functional/plugin/lsp_spec.lua (hunk 6 above)
Comment (gbrlsnchs) [reply test/functional/plugin/lsp_spec.lua:1434]: Done, that is a test case now, you can find it by reading the comments I left in the tests. :smile_cat:

File: runtime/lua/vim/lsp/util.lua (hunk 7 above)
Comment (gbrlsnchs) [reply runtime/lua/vim/lsp/util.lua:544]: You know what, it was weird to have that function declared inside another one. Now it's a single, private function that checks whether item should be added to tagstack. :smiley:

File: runtime/lua/vim/lsp/util.lua (hunk 5 above)
Comment (tjdevries) [reply runtime/lua/vim/lsp/util.lua:557]: <3 That's what I was wondering. Thanks for following up on this!

File: test/functional/plugin/lsp_spec.lua (hunk 6 above)
Comment (tjdevries) [reply test/functional/plugin/lsp_spec.lua:1446]: So another test case I'd like to see is that you do something with `location` where you set a new `uri` to go to. 

You may have to extend: https://github.com/neovim/neovim/pull/12262/files#diff-2a259d96bc36492344afe33e6d418d83R1361-R1369 to allow for a new `uri` and to have it opened with text inside of it in this section:  https://github.com/neovim/neovim/pull/12262/files#diff-2a259d96bc36492344afe33e6d418d83R1352-R1359

So basically, add another `uri`, like `fake/uri_2` or similar, and then jump there and back.

Also, using something like using `<C-t>` would be pretty cool here to see see if it actually jumps back to where you were before.

**Review by tjdevries — CHANGES_REQUESTED** [review]
I think we can add a test case that "proves" that the buffer stuff is working by extending the tests a bit more. I gave you an example.

Additionally, you should probably do something like separate the tests a bit more with more `it(..., function()...end)` stuff, so that you can pinpoint more easily _exactly_ what failed if something fails in the future.

Anyway, thanks for doing this!

File: test/functional/plugin/lsp_spec.lua (hunk 6 above)
Comment (gbrlsnchs) [reply test/functional/plugin/lsp_spec.lua:1446]: Good point, I'd love to organize the whole test.

> So basically, add another uri, like fake/uri_2 or similar, and then jump there and back.

I tried that but unfortunately the test threw an error saying it couldn't switch to another buffer. :crying_cat_face: 

Basically, inside a `test` context, I can only call `nvim_set_current_buf` once. If I try to call it again (so I can jump between buffers), it throws that error. I have no clue whether I'm doing something wrong or it's a limitation of the test itself.

> Also, using something like using <C-t> would be pretty cool here to see see if it actually jumps back to where you were before.

This makes total sense, we need to prove whether we're properly pushing items to the tagstack. I need two things to make this possible:
- Solve the issue I cited above
- Learn how to trigger `<C-t>` (is it via `execute normal!`?)

**Comment by gbrlsnchs** [comment]
> Additionally, you should probably do something like separate the tests a bit more with more it(..., function()...end) stuff, so that you can pinpoint more easily exactly what failed if something fails in the future.

I agree. I had trouble finding out what was failing while developing this. As I pointed in my other comment, I'll tidy the whole test a little bit.

File: test/functional/plugin/lsp_spec.lua (hunk 6 above)
Comment (tjdevries) [reply test/functional/plugin/lsp_spec.lua:1446]: Hmm, it could be because you didn't have another buffer loaded. That's why I think you need to load it in the `before_each` section. Does that make sense?

As for the trigger the key sequence, you can use `helpers.feed`, which should work fine.

Comment (gbrlsnchs) [reply test/functional/plugin/lsp_spec.lua:1446]: I tried loading the new buffer using the same code from the `before_each` hook just before the call to `nvim_set_current_buf` (inside the `test` section), but I had no luck. I think I have even loaded it in the `before_each` hook, but still no success.

I'll try it again later and if I have no success, I'll ask for help here or on IRC. Thanks again!

**Comment by tjdevries** [comment]
(Let me know if you need help tidying up the tests or anything like that)

**Comment by gbrlsnchs** [comment]
> (Let me know if you need help tidying up the tests or anything like that)

Yeah, I just resumed doing this PR and... I have hit a wall. I'm trying to run the following code in tests:
```lua
  describe('lsp.util.jump_to_location', function()

-- ...

    local create_buf = function(uri)
      return exec_lua([[
        local bufnr = vim.uri_to_bufnr(...)
        local lines = {"1st line of text", "å å ɧ 汉语 ↥ 🤦 🦄"}
        vim.api.nvim_buf_set_lines(bufnr, 0, 1, false, lines)
        return bufnr
      ]], uri)
    end

    local map_filenames = function(fn, arr)
      local bufnrs = {}
      for i, v in ipairs(arr) do
        bufnrs[i] = fn(v)
      end
      return bufnrs
    end

    local fake_location = function(uri, start_line, start_char, end_line, end_char)
      return {
        uri = uri,
        range = {
          start = { line = start_line, character = start_char },
          ["end"] = { line = end_line, character = end_char },
        },
      }
    end

    local jump = function(msg)
      eq(true, exec_lua('return vim.lsp.util.jump_to_location(...)', msg))
      eq(target_bufnr, exec_lua[[return vim.fn.bufnr('%')]])
      return {
        line = exec_lua[[return vim.fn.line('.')]],
        col = exec_lua[[return vim.fn.col('.')]],
      }
    end

    it('pushes to tagstack when jumping from one buffer to another', function()
      local filenames = {
        'file://first/buffer',
        'file://second/buffer',
      }
      local bufnrs = map_filenames(create_buf, filenames)
      local stack = {}

      -- Prepare the test by manually moving to the first buffer.
      exec_lua([[
        vim.api.nvim_set_current_buf(...)
        vim.api.nvim_win_set_cursor(0, {1, 0})
      ]], bufnrs[1])

      -- Jump to line #1, col #1 (tagname is '1st') of buffer #2.
      jump(fake_location(filenames[2], 0, 0, 0, 0))
      stack = exec_lua[[return vim.fn.gettagstack()]]
      eq(1, stack.length)
      -- Cursor came from line #1, col #1 of buffer #1.
      eq({1, 1, 1, 0}, stack.items[1].from)
    end)

-- ...

end)
```

But the test fails whenever I call `nvim_set_current_buf` more than once. In this case, I first move to `bufnrs[1]`, which is fine, but when I call `jump`, it invokes `vim.lsp.util.jump_to_location`, which internally calls `nvim_set_current_buf` again. That's the point where the test fails with the following message: `Failed to switch to buffer 4`, where `4` is `bufnrs[2]`.

**Comment by tjdevries** [comment]
I think you need to be doing the buffer mapping stuff in `exec_lua`. I don't think those are actually opened for you.

If that's not it, ping me again and I'll look more.

**Comment by gbrlsnchs** [comment]
@tjdevries After some delay (sorry about that), I might have finished this PR. The reason I couldn't jump from one buffer to another in tests was because `hidden` was not set, so the jump triggered an error.

This PR might need some eventual polish, but tests are there! Let me know if I should make any changes! :smile_cat:

**Comment by tjdevries** [comment]
I think this LGTM :) Thanks for adding the tests!

@mjlbach do you agree?

**Comment by mjlbach** [comment]
+1, just read through and the changes LGTM

**Review by clason — APPROVED** [review]

**Comment by tjdevries** [comment]
hehe, looks like everyone approved and I didn't click any buttons. Merging

**Comment by fsouza** [comment]
👋 all! With this change, I'm observing a weird behavior when jump to definitions within the same file, in that case the stack isn't updated and if I press c-o after jumping to definition. For example, take the simple code below:

```python
x = 10
y = 11

print(x + y)
```

If I'm with the cursor under `y` in line 4 and jump to definition, that correctly takes me to line 2, but then if I press c-o I move to the first line (or whatever the previous position was before going to `y`).

Is this the expected behavior? Should I open a new issue?

**Comment by mjlbach** [comment]
No, that doesn't seem like correct behavior. You should open an issue.

**Comment by fsouza** [comment]
> No, that doesn't seem like correct behavior. You should open an issue.

opened #14446, will come back later today or tomorrow to add some tests.

**Comment by gbrlsnchs** [comment]
Yeah, I removed the `normal! m'"` part naively thinking it wouldn't be used anymore since you already could jump back using the tagstack, however now I notice it's different since with the jumplist you can go back and forth, which is not the case with the tagstack. Sorry for that! :disappointed_relieved: I only focused on not breaking existing tests, and this feature was not being tested. :disappointed: 

@fsouza Still, I think the fix can leverage this PR by only adding the position to the jumplist under the same circumstances it would be added to the tagstack, avoiding consecutive repeated positions (I think you already did that in your PR).

Thanks for pointing that out and trying to fix that!
